File: browser_ai_gui/protocol.py
"""
Browser.AI WebSocket Protocol

Shared protocol definitions for communication between the Chrome extension
and the Browser.AI server over WebSocket connections.

Namespace: /extension

This module defines the data structures and message formats used in the
WebSocket communication protocol to ensure consistency between the Python
server and TypeScript extension client.
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional, Union


# ============================================================================
# Enums
# ============================================================================


class LogLevel(Enum):
    """Log levels for events"""

    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    RESULT = "RESULT"


class EventType(Enum):
    """Types of events that can be captured"""

    LOG = "log"
    AGENT_START = "agent_start"
    AGENT_STEP = "agent_step"
    AGENT_ACTION = "agent_action"
    AGENT_RESULT = "agent_result"
    AGENT_COMPLETE = "agent_complete"
    AGENT_ERROR = "agent_error"
    AGENT_PAUSE = "agent_pause"
    AGENT_RESUME = "agent_resume"
    AGENT_STOP = "agent_stop"
    USER_HELP_NEEDED = "user_help_needed"


# ============================================================================
# Data Structures
# ============================================================================


@dataclass
class LogEvent:
    """
    Log event broadcast from server to clients

    Attributes:
        timestamp: ISO 8601 formatted timestamp
        level: Log level
        logger_name: Name of the logger that generated this event
        message: Log message content
        event_type: Type of event
        metadata: Optional additional metadata
    """

    timestamp: str
    level: str  # LogLevel value
    logger_name: str
    message: str
    event_type: str  # EventType value
    metadata: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            "timestamp": self.timestamp,
            "level": self.level,
            "logger_name": self.logger_name,
            "message": self.message,
            "event_type": self.event_type,
            "metadata": self.metadata or {},
        }


@dataclass
class TaskStatus:
    """
    Task status information

    Attributes:
        is_running: Whether a task is currently running
        current_task: Description of current task (None if no task)
        has_agent: Whether an agent instance exists
        is_paused: Whether the task is paused
        cdp_endpoint: CDP WebSocket endpoint being used
    """

    is_running: bool
    current_task: Optional[str]
    has_agent: bool
    is_paused: Optional[bool] = None
    cdp_endpoint: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        return {
            "is_running": self.is_running,
            "current_task": self.current_task,
            "has_agent": self.has_agent,
            "is_paused": self.is_paused,
            "cdp_endpoint": self.cdp_endpoint,
        }


@dataclass
class ActionResult:
    """
    Generic action result

    Attributes:
        success: Whether the action succeeded
        message: Optional success message
        error: Optional error message
    """

    success: bool
    message: Optional[str] = None
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        result = {"success": self.success}
        if self.message:
            result["message"] = self.message
        if self.error:
            result["error"] = self.error
        return result


# ============================================================================
# Client -> Server Event Payloads
# ============================================================================


@dataclass
class StartTaskPayload:
    """
    Event: start_task
    Request to start a new Browser.AI task

    Attributes:
        task: Task description
        cdp_endpoint: Optional CDP WebSocket endpoint
        is_extension: Flag to indicate extension mode
    """

    task: str
    cdp_endpoint: Optional[str] = None
    is_extension: Optional[bool] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "StartTaskPayload":
        """Create from dictionary"""
        return cls(
            task=data.get("task", ""),
            cdp_endpoint=data.get("cdp_endpoint"),
            is_extension=data.get("is_extension"),
        )


# ============================================================================
# Server -> Client Event Payloads
# ============================================================================


@dataclass
class TaskStartedPayload:
    """
    Event: task_started
    Confirmation that task has started

    Attributes:
        message: Status message
        success: Whether task started successfully
        error: Optional error message
    """

    message: str
    success: Optional[bool] = None
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        result = {"message": self.message}
        if self.success is not None:
            result["success"] = self.success
        if self.error:
            result["error"] = self.error
        return result


@dataclass
class ErrorPayload:
    """
    Event: error
    Error message from server

    Attributes:
        message: Error message
        details: Optional additional details
    """

    message: str
    details: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization"""
        result = {"message": self.message}
        if self.details:
            result["details"] = self.details
        return result


# ============================================================================
# Protocol Constants
# ============================================================================

WEBSOCKET_NAMESPACE = "/extension"
DEFAULT_SERVER_URL = "http://localhost:5000"
MAX_RECONNECTION_ATTEMPTS = 5
RECONNECTION_DELAY_MS = 1000
MAX_LOGS = 1000


# ============================================================================
# Helper Functions
# ============================================================================


def serialize_log_event(
    timestamp: datetime,
    level: Union[LogLevel, str],
    logger_name: str,
    message: str,
    event_type: Union[EventType, str],
    metadata: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """
    Serialize a log event for transmission

    Args:
        timestamp: Event timestamp
        level: Log level (enum or string)
        logger_name: Logger name
        message: Log message
        event_type: Event type (enum or string)
        metadata: Optional metadata

    Returns:
        Dictionary ready for JSON serialization
    """
    log_event = LogEvent(
        timestamp=(
            timestamp.isoformat() if isinstance(timestamp, datetime) else timestamp
        ),
        level=level.value if isinstance(level, LogLevel) else level,
        logger_name=logger_name,
        message=message,
        event_type=(
            event_type.value if isinstance(event_type, EventType) else event_type
        ),
        metadata=metadata,
    )
    return log_event.to_dict()


def create_task_status(
    is_running: bool,
    current_task: Optional[str] = None,
    has_agent: bool = False,
    is_paused: Optional[bool] = None,
    cdp_endpoint: Optional[str] = None,
) -> TaskStatus:
    """
    Create a task status object

    Args:
        is_running: Whether a task is running
        current_task: Current task description
        has_agent: Whether an agent exists
        is_paused: Whether the task is paused
        cdp_endpoint: CDP endpoint being used

    Returns:
        TaskStatus object
    """
    return TaskStatus(
        is_running=is_running,
        current_task=current_task,
        has_agent=has_agent,
        is_paused=is_paused,
        cdp_endpoint=cdp_endpoint,
    )


def create_action_result(
    success: bool, message: Optional[str] = None, error: Optional[str] = None
) -> ActionResult:
    """
    Create an action result object

    Args:
        success: Whether the action succeeded
        message: Optional success message
        error: Optional error message

    Returns:
        ActionResult object
    """
    return ActionResult(success=success, message=message, error=error)


def create_error_payload(message: str, details: Optional[str] = None) -> ErrorPayload:
    """
    Create an error payload object

    Args:
        message: Error message
        details: Optional additional details

    Returns:
        ErrorPayload object
    """
    return ErrorPayload(message=message, details=details)

File: browser_ai_extension/browse_ai/src/types/protocol.ts
/**
 * Browser.AI WebSocket Protocol
 *
 * Shared protocol definitions for communication between the Chrome extension
 * and the Browser.AI server over WebSocket connections.
 *
 * Namespace: /extension
 */

// ============================================================================
// Enums
// ============================================================================

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARNING = 'WARNING',
  ERROR = 'ERROR',
  RESULT = 'RESULT',
}

export enum EventType {
  LOG = 'log',
  AGENT_START = 'agent_start',
  AGENT_STEP = 'agent_step',
  AGENT_ACTION = 'agent_action',
  AGENT_RESULT = 'agent_result',
  AGENT_COMPLETE = 'agent_complete',
  AGENT_ERROR = 'agent_error',
  AGENT_PAUSE = 'agent_pause',
  AGENT_RESUME = 'agent_resume',
  AGENT_STOP = 'agent_stop',
  USER_HELP_NEEDED = 'user_help_needed',
}

// ============================================================================
// Settings & Configuration
// ============================================================================

/**
 * Extension settings stored in Chrome storage
 */
export interface ExtensionSettings {
  serverUrl: string
  devMode: boolean
  autoReconnect: boolean
  maxLogs: number
  showNotifications: boolean
}

export const DEFAULT_SETTINGS: ExtensionSettings = {
  serverUrl: 'http://localhost:5000',
  devMode: false,
  autoReconnect: true,
  maxLogs: 1000,
  showNotifications: true,
}

// ============================================================================
// Data Structures
// ============================================================================

/**
 * Log event broadcast from server to clients
 */
export interface LogEvent {
  timestamp: string // ISO 8601 format
  level: LogLevel | string
  logger_name: string
  message: string
  event_type: EventType | string
  metadata?: Record<string, any>
}

/**
 * Task status information
 */
export interface TaskStatus {
  is_running: boolean
  current_task: string | null
  has_agent: boolean
  is_paused?: boolean
  cdp_endpoint?: string
}

/**
 * Generic action result
 */
export interface ActionResult {
  success: boolean
  message?: string
  error?: string
}

// ============================================================================
// Client -> Server Events (Emitted by Extension)
// ============================================================================

/**
 * Event: extension_connect
 * Sent when extension client connects to establish session
 * Payload: none
 */
export type ExtensionConnectEvent = void

/**
 * Event: start_task
 * Request to start a new Browser.AI task
 */
export interface StartTaskPayload {
  task: string // Task description
  cdp_endpoint?: string // Optional CDP WebSocket endpoint
  is_extension?: boolean // Flag to indicate extension mode
}

/**
 * Event: stop_task
 * Request to stop the currently running task
 * Payload: none
 */
export type StopTaskEvent = void

/**
 * Event: pause_task
 * Request to pause the currently running task
 * Payload: none
 */
export type PauseTaskEvent = void

/**
 * Event: resume_task
 * Request to resume a paused task
 * Payload: none
 */
export type ResumeTaskEvent = void

/**
 * Event: get_status
 * Request current task status
 * Payload: none
 */
export type GetStatusEvent = void

// ============================================================================
// Server -> Client Events (Received by Extension)
// ============================================================================

/**
 * Event: status
 * Broadcast of current task status (sent on connect and status changes)
 */
export type StatusEvent = TaskStatus

/**
 * Event: log_event
 * Real-time log event from the Browser.AI agent
 */
export type LogEventBroadcast = LogEvent

/**
 * Event: task_started
 * Confirmation that task has started
 */
export interface TaskStartedPayload {
  message: string
  success?: boolean
  error?: string
}

/**
 * Event: task_action_result
 * Result of a task action (stop, pause, resume)
 */
export type TaskActionResultPayload = ActionResult

/**
 * Event: error
 * Error message from server
 */
export interface ErrorPayload {
  message: string
  details?: string
}

// ============================================================================
// Socket Event Map (for type-safe socket.io usage)
// ============================================================================

/**
 * Client to Server event map
 */
export interface ClientToServerEvents {
  extension_connect: () => void
  start_task: (data: StartTaskPayload) => void
  stop_task: () => void
  pause_task: () => void
  resume_task: () => void
  get_status: () => void
}

/**
 * Server to Client event map
 */
export interface ServerToClientEvents {
  status: (status: TaskStatus) => void
  log_event: (event: LogEvent) => void
  task_started: (data: TaskStartedPayload) => void
  task_action_result: (result: ActionResult) => void
  task_result: (result: { task: string; success: boolean; history?: string; agent_id?: string }) => void
  error: (error: ErrorPayload) => void
  structured_event: (event: any) => void // New structured event system
  connect: () => void
  disconnect: () => void
}

// ============================================================================
// Helper Types
// ============================================================================

/**
 * Type-safe socket type for extension
 */
export type ExtensionSocket = {
  on<Event extends keyof ServerToClientEvents>(
    event: Event,
    listener: ServerToClientEvents[Event],
  ): void
  emit<Event extends keyof ClientToServerEvents>(
    event: Event,
    ...args: Parameters<ClientToServerEvents[Event]>
  ): void
  connected: boolean
  close(): void
}

// ============================================================================
// Constants
// ============================================================================

export const WEBSOCKET_NAMESPACE = '/extension'
export const DEFAULT_SERVER_URL = 'http://localhost:5000'
export const MAX_RECONNECTION_ATTEMPTS = 5
export const RECONNECTION_DELAY_MS = 1000
export const MAX_LOGS = 1000

// ============================================================================
// Validation Helpers
// ============================================================================

export function isLogEvent(obj: any): obj is LogEvent {
  return (
    obj &&
    typeof obj.timestamp === 'string' &&
    typeof obj.level === 'string' &&
    typeof obj.logger_name === 'string' &&
    typeof obj.message === 'string' &&
    typeof obj.event_type === 'string'
  )
}

export function isTaskStatus(obj: any): obj is TaskStatus {
  return (
    obj &&
    typeof obj.is_running === 'boolean' &&
    (obj.current_task === null || typeof obj.current_task === 'string') &&
    typeof obj.has_agent === 'boolean'
  )
}

export function isActionResult(obj: any): obj is ActionResult {
  return obj && typeof obj.success === 'boolean'
}
